app/src/services/cpk/cpk.ts

```diff
@@ -267,230 +262,31 @@ class CPKService {
-      const network = await this.provider.getNetwork()
-      const networkId = network.chainId
-
-      const transactions: Transaction[] = []
-
-      const txOptions: TxOptions = {}
-      await this.getGas(txOptions)
-
-      const buyAmount = await this.subRelayFee(amount)
-
-      let collateralAddress
-
-      if (collateral.address === pseudoNativeAssetAddress) {
-        // ultimately WETH will be the collateral if we fund with native ether
-        collateralAddress = getWrapToken(networkId).address
-
-        // we need to send the funding amount in native ether
-        if (!this.isSafeApp) {
-          txOptions.value = buyAmount
-        }
-
-        // Step 0: Wrap ether
-        transactions.push({
-          to: collateralAddress,
-          value: buyAmount.toString(),
-        })
-      } else {
-        collateralAddress = await marketMaker.getCollateralToken()
-      }
-
-      const marketMakerAddress = marketMaker.address
-      const collateralService = new ERC20Service(this.provider, account, collateralAddress)
-      logger.log(`CPK address: ${this.cpk.address}`)
-      const outcomeTokensToBuy = await marketMaker.calcBuyAmount(buyAmount, outcomeIndex)
-      logger.log(`Min outcome tokens to buy: ${outcomeTokensToBuy}`)
-
-      // Check  if the allowance of the CPK to the market maker is enough.
-      const hasCPKEnoughAlowance = await collateralService.hasEnoughAllowance(
-        this.cpk.address,
-        marketMakerAddress,
-        buyAmount,
-      )
-
-      if (!hasCPKEnoughAlowance) {
-        // Step 1:  Approve unlimited amount to be transferred to the market maker)
-        transactions.push({
-          to: collateralAddress,
-          data: ERC20Service.encodeApproveUnlimited(marketMakerAddress),
-        })
-      }
-
-      // Step 2: Transfer the amount of collateral being spent from the user to the CPK
-      // If we are funding with native ether we can skip this step
-      // If we are signed in as a safe we don't need to transfer
-      if (!this.isSafeApp && collateral.address !== pseudoNativeAssetAddress) {
-        // Step 2: Transfer the amount of collateral being spent from the user to the CPK
-        transactions.push({
-          to: collateralAddress,
-          data: ERC20Service.encodeTransferFrom(account, this.cpk.address, buyAmount),
-        })
-      }
-      // Step 3: Buy outcome tokens with the CPK
-      transactions.push({
-        to: marketMakerAddress,
-        data: MarketMakerService.encodeBuy(buyAmount, outcomeIndex, outcomeTokensToBuy),
-      })
+  service = (params: any) => ({ ...params, service: this })
 
+  buyOutcomes = async (params: CPKBuyOutcomesParams): Promise<TransactionReceipt> => {
+    try {
+      const { transactions, txOptions } = await pipe(setup, fee, wrap, approve, transfer, buy)(this.service(params))
+      const { setTxHash, setTxState } = params
       return this.execTransactions(transactions, txOptions, setTxHash, setTxState)
```

Comment:
Im wondering would it make sense to make execTransactions one pipe function as well since it will be always repeated for every cpk function?

Reply:
great idea! will try it out

Reply:
![Screenshot 2021-07-06 at 14 32 04](https://user-images.githubusercontent.com/33226956/124600382-f007b700-de66-11eb-808f-4566050e827d.png)
Very classy solution I must say :D